File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <sstream>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> uidToNickname;
    vector<pair<string, string>> logs; 

    for (const string& line : record) {
        stringstream ss(line);
        string action, uid, nickname;
        ss >> action >> uid;

        if (action == "Enter" || action == "Change") {
            ss >> nickname;
            uidToNickname[uid] = nickname; 
        }

        if (action == "Enter" || action == "Leave") {
            logs.push_back({action, uid}); 
        }
    }

    for (const auto& [action, uid] : logs) {
        string nickname = uidToNickname[uid];
        if (action == "Enter") {
            answer.push_back(nickname + "님이 들어왔습니다.");
        } else if (action == "Leave") {
            answer.push_back(nickname + "님이 나갔습니다.");
        }
    }

    return answer;
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
  uid_to_nick = {}
  actions = []

  for rec in record:
      parts = rec.split()
      if parts[0] == "Enter":
          uid_to_nick[parts[1]] = parts[2]
          actions.append((parts[0], parts[1]))
      elif parts[0] == "Leave":
          actions.append((parts[0], parts[1]))
      elif parts[0] == "Change":
          uid_to_nick[parts[1]] = parts[2]

  result = []
  for action, uid in actions:
      nickname = uid_to_nick[uid]
      if action == "Enter":
          result.append(f"{nickname}님이 들어왔습니다.")
      elif action == "Leave":
          result.append(f"{nickname}님이 나갔습니다.")

  return result